### DISCUSSION TITLE
Python User Defined Aggregation Functions (UDAF) with combiners

### DISCUSSION THREAD

**Discussion by ogrisel** [duckdb/duckdb#5117] [General]
The map method introduced in #1569 opens a lot of flexibility to inject arbitrary Python logic into the optimized query plan with builtin parallelization. However it's limiting to one-to-one (vector) functions.

Do you have any plan to extend this with support for (vector) Python UDAFs? Ideally with the possibility to specify a main reducer/aggregator vector function + a related combiner to make it possible to parallelize the execution of the UDAFs. E.g.:

```python
from __future__ import annotations
import pandas as pd
from dataclasses import dataclass


@dataclass
class MeanAggregation:
    partial_sum: pd.Series
    partial_count: pd.Series

    @staticmethod
    def reduce(df: pd.DataFrame) -> MeanAggregation:
        return MeanAggregation(df.sum(skipna=True), df.count())

    def combine(self, other: MeanAggregation) -> MeanAggregation:
        return MeanAggregation(
            self.partial_sum + other.partial_sum,
            self.partial_count + self.partial_count,
        )

    def finalize(self) -> pd.Series:
        return self.partial_sum / self.partial_count
```

Here is a code snippet to show how it can be used to compute the aggregation in parallel:

```python
>>> import numpy as np
>>> from concurrent.futures import ThreadPoolExecutor
>>> big_dataframe = pd.DataFrame(
...     {
...         "a": np.random.randn(int(1e8)),
...         "b": np.random.randint(0, 10, int(1e8)),
...     }
... )
>>> chunked_dataframes = np.array_split(big_dataframe, 100)
>>> final_aggregation = None
>>> tpe = ThreadPoolExecutor()
>>> for agg in tpe.map(MeanAggregation.reduce, chunked_dataframes):
...     if final_aggregation is None:
...         final_aggregation = agg
...     else:
...         final_aggregation.combine(agg)
... 
>>> final_aggregation.finalize()
a    0.001118
b    4.500636
dtype: float64
>>> big_dataframe.mean()
a   -0.000073
b    4.499862
dtype: float64
```

Note: arithmetic operations do not happen in the same order when working with on chunked data, hence the slightly different results caused by `float64` rounding errors. The above code is quite naive but there would be a way to make it more numerically stable but I did not do it for the sake of simplicity.

Also here the combiner calls happen sequentially in the main thread but ideally in duckdb, the combiner calls would already start right away in the workers to keep the memory usage under control.

Note, I opened a similar issue for the Ibis project:
- https://github.com/ibis-project/ibis/issues/4707

**Comment by hannes** [comment]
See here for some background: https://github.com/duckdb/duckdb/discussions/4797#discussioncomment-3726080

**Comment by ogrisel** [comment]
Instead of asking the user to provide a (data)class, it would be possible to expose an API that accepts a triplet of Python callables (equivalent to the `reduce`, `combine` and `finalize` methods above) using a tuple instead of a dataclass for the intermediate datastructure as done in the Dask Dataframe API: https://docs.dask.org/en/latest/dataframe-groupby.html#aggregate
